image/unpacker.go

```diff
@@ -0,0 +1,238 @@
+// Copyright 2016 The Linux Foundation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package image
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/opencontainers/image-spec/schema"
+	"github.com/pkg/errors"
+)
+
+// Unpacker is a structure that encapsulates the image layout unpacking
+// functionality.
+type Unpacker struct {
+	// If set, the ownership of files, directories and symblic links defined in
+	// the layers, will be preserved in the unpacked root file system.
+	PreserveOwnership bool
```

Comment:
Might it be better that this takes a function that translates the entries, rather than just a bool? There are a lot of different ways to do it and make this work with this bool may be a little awkward.

`PreserveOwnership`, the default (I guess) would this:

```
func PreserveOwnship(uid, gid string) (string, string, error) {
  u, err := os/user.Current()
  if err { return "", "", err }

  return u.Uid, u.Gid, nil
}
```

Result of this function can be memoized to minimize the number of calls during the unpacking process.